tests/test_json_schema.py

```diff
@@ -2692,73 +2683,202 @@ class Model(BaseModel):
+        pet_type: Literal['reptile', 'lizard']
+
+    class Model(BaseModel):
+        pet: Annotated[Union[Cat, Dog, Lizard], Field(..., discriminator='pet_type')]
+
+    assert Model.model_json_schema() == {
+        '$defs': {
+            'Cat': {
+                'properties': {'pet_type': {'const': 'cat', 'title': 'Pet Type'}},
+                'required': ['pet_type'],
+                'title': 'Cat',
+                'type': 'object',
+            },
+            'Dog': {
+                'properties': {'pet_type': {'const': 'dog', 'title': 'Pet Type'}},
+                'required': ['pet_type'],
+                'title': 'Dog',
+                'type': 'object',
+            },
+            'Lizard': {
+                'properties': {'pet_type': {'enum': ['reptile', 'lizard'], 'title': 'Pet Type'}},
+                'required': ['pet_type'],
+                'title': 'Lizard',
+                'type': 'object',
+            },
+        },
         'properties': {
             'pet': {
-                'title': 'Pet',
                 'discriminator': {
-                    'propertyName': 'pet_type',
                     'mapping': {
                         'cat': '#/$defs/Cat',
                         'dog': '#/$defs/Dog',
-                        'reptile': '#/$defs/Lizard',
                         'lizard': '#/$defs/Lizard',
+                        'reptile': '#/$defs/Lizard',
                     },
+                    'propertyName': 'pet_type',
                 },
                 'oneOf': [
                     {'$ref': '#/$defs/Cat'},
                     {'$ref': '#/$defs/Dog'},
                     {'$ref': '#/$defs/Lizard'},
+                    {'$ref': '#/$defs/Lizard'},
                 ],
+                'title': 'Pet',
             }
         },
         'required': ['pet'],
+        'title': 'Model',
+        'type': 'object',
+    }
+
+
+def test_nested_discriminated_union():
+    class BlackCatWithHeight(BaseModel):
+        color: Literal['black']
+        info: Literal['height']
+        height: float
+
+    class BlackCatWithWeight(BaseModel):
+        color: Literal['black']
+        info: Literal['weight']
+        weight: float
+
+    BlackCat = Annotated[Union[BlackCatWithHeight, BlackCatWithWeight], Field(discriminator='info')]
+
+    class WhiteCat(BaseModel):
+        color: Literal['white']
+        white_cat_info: str
+
+    class Cat(BaseModel):
+        pet: Annotated[Union[BlackCat, WhiteCat], Field(discriminator='color')]
+
+    assert Cat.model_json_schema() == {
         '$defs': {
-            'BlackCat': {
-                'title': 'BlackCat',
+            'BlackCatWithHeight': {
+                'properties': {
+                    'color': {'const': 'black', 'title': 'Color'},
+                    'height': {'title': 'Height', 'type': 'number'},
+                    'info': {'const': 'height', 'title': 'Info'},
+                },
+                'required': ['color', 'info', 'height'],
```

Comment:
(For more complex schemas I think it's even more attractive to say these would normally be a "level-up" in the schema so that they're not repeated each time...)